vpr/src/place/timing/PlacerSetupSlacks.h

```diff
@@ -0,0 +1,120 @@
+
+#pragma once
+
+#include "vtr_vec_id_set.h"
+#include "timing_info_fwd.h"
+#include "clustered_netlist_utils.h"
+#include "place_delay_model.h"
+#include "vpr_net_pins_matrix.h"
+
+/**
+ * @brief PlacerSetupSlacks returns the RAW setup slacks of clustered netlist connection.
+ *
+ * Usage
+ * =====
+ * This class mirrors PlacerCriticalities by both its methods and its members. The only
+ * difference is that this class deals with RAW setup slacks returned by SetupTimingInfo
+ * rather than criticalities. See the documentation on PlacerCriticalities for more.
+ *
+ * RAW setup slacks are unlike criticalities. Their values are not confined between
+ * 0 and 1. Their values can be either positive or negative.
+ *
+ * This class also provides iterating over the clustered netlist connections/pins that
+ * have modified setup slacks by the last call to update_setup_slacks(). However, this
+ * utility is mainly used for incrementally committing the setup slack values into the
+ * structure `connection_setup_slack` used by many placer routines.
+ */
+class PlacerSetupSlacks {
+  public: //Types
+    typedef vtr::vec_id_set<ClusterPinId>::iterator pin_iterator;
+    typedef vtr::vec_id_set<ClusterNetId>::iterator net_iterator;
+
+    typedef vtr::Range<pin_iterator> pin_range;
+    typedef vtr::Range<net_iterator> net_range;
+
+  public: //Lifetime
+    ///@brief Allocates space for the timing_place_setup_slacks_ data structure.
+    PlacerSetupSlacks(const ClusteredNetlist& clb_nlist,
+                      const ClusteredPinAtomPinsLookup& netlist_pin_lookup,
+                      std::shared_ptr<const SetupTimingInfo> timing_info);
+
+    PlacerSetupSlacks(const PlacerSetupSlacks& clb_nlist) = delete;
+    PlacerSetupSlacks& operator=(const PlacerSetupSlacks& clb_nlist) = delete;
+
+  public: //Accessors
+    ///@brief Returns the setup slack of the specified connection.
+    float setup_slack(ClusterNetId net, int ipin) const { return timing_place_setup_slacks_[net][ipin]; }
+
+    /**
+     * @brief Returns the range of clustered netlist pins (i.e. ClusterPinIds)
+     *        which were modified by the last call to PlacerSetupSlacks::update_setup_slacks().
+     */
+    pin_range pins_with_modified_setup_slack() const;
+
+  public: //Modifiers
+    /**
+     * @brief Updates setup slacks based on the atom netlist setup slacks provided
+     *        by timing_info_.
+     *
+     * Should consistently call this method after the most recent timing analysis to
+     * keep the setup slacks stored in this class in sync with the timing analyzer.
+     * If out of sync, then the setup slacks cannot be incrementally updated on
```

Comment:
updated on -> updated

Reply:
Merge in the extra information from the .cpp file comment:

 * @brief Updated the setup slacks in the timing_place_setup_slacks_ data structure.
 *
 * If the setup slacks are not updated immediately after each time we call
 * timing_info->update(), then timing_info->pins_with_modified_setup_slack()
 * cannot accurately account for all the pins that need to be updated.
 *
 * In this case, `recompute_required` would be true, and we update all setup slacks
 * from scratch.